### ISSUE TITLE
FALCON can't handle valid fasta file

### ISSUE THREAD

**Issue by afinit** [PacificBiosciences/FALCON#224]
I have selected a subset of PacBio reads to do assemblies of specific features of my genome. When I do this I set all of my fasta sequences to one line for the ID and one line for the sequence because it's easier to work with on the command line. While this is a valid fasta file, FALCON throws the following error:

File subset_falcon_reads.fasta, Line 40: Fasta line is too long (> 9998 chars)

I'm not sure I understand the reason for a limit here.

**Comment by flowers9** [comment]
It's just a coding choice, from what I can see.  It was a little work, but I modified my version to handle indefinite length lines, and it doesn't seem to cause any issues.  However, I had code to do it at hand from other projects, it would probably be a fair bit of work to do from scratch.

It shouldn't be too hard to modify your fasta files to wrap the sequence at some length less than 9998.  Just make sure not to wrap the header lines. ;)

**Comment by pb-cdunn** [comment]
@flowers9, We'd be happy to apply your patch to `fasta2DB.c`. If possible, send us a pull-request on **PacificBiosciences/DAZZ_DB**.

**Comment by pb-cdunn** [comment]
Really, this would be a welcome change to our version of daligner. Could you show us your changes somehow? Maybe store your version of daligner in GitHub as a private repo and add us as a collaborator. Or paste your code into a GitHub "gist".

**Comment by dshvets** [comment]
I have this same problem when I am trying to run Falcon with Pacbio data. Has this been fixed yet? Or can someone post how they got around this?

**Comment by pb-cdunn** [comment]
You can definitely work around this by reformatting your .fasta input. Some people restrict line-lengths to 60 bps. You can use this (if you have **pbcore**):
```
from pbcore.io import FastaIO

def linewrap_fasta(ifn, ofn):
    with FastaIO.FastaReader(ifn) as fa_in:
        with FastaIO.FastaWriter(ofn) as fa_out:
            for rec in fa_in:
                fa_out.writeRecord(rec)
```
If we can't get daligner updated to handle this, we'll add a pre-processor to FALCON that doesn't depend on **pbcore** (since this wrapping is simple).

**Comment by dshvets** [comment]
I'm sorry I am still confused, where exactly am I supposed to add this code? Or am I supposed to edit my fasta file before I run the script?
I am sorry I am not very familiar with all of the terminology, I have simply downloaded your tool, successfully run the test set but when I try to run it with my pacbio data it is giving me the same error as the person above. Thank you.

**Comment by pb-cdunn** [comment]
Yes, you would have to alter your .fasta file before starting FALCON. And you would need to install **pbcore** to use that script.

In the meantime, I'll work on something to make that automatic, to save you the trouble...

**Comment by dshvets** [comment]
Ok I will try this - thank you for your help.

**Comment by pb-cdunn** [comment]
If you can pull the `fasta-wrap` branch, you can try this.
* https://github.com/PacificBiosciences/FALCON/tree/fasta-wrap

I have a to a bit of testing before making this official.

**Comment by pb-cdunn** [comment]
Update. We will take the change PacificBiosciences/DAZZ_DB#3 from @flowers9 to fix this in fasta2DB, but not til we test it on Tuesday. If you cannot wait Til Tuesday, ["hush hush"](http://www.azlyrics.com/lyrics/tiltuesday/voicescarry.html), or try the temporary fix on the fasta-wrap branch.

**Comment by dshvets** [comment]
I can wait :) Thank you for letting me know!

**Comment by aliceb5** [comment]
Hi all,  

I'm hitting this ">9998 character" error right now too. Will you post a message when there an updated version of Falcon is available to try?

Thanks!

**Comment by pb-cdunn** [comment]
If you pull the latest PacificBiosciences/DAZZ_DB, you'll have it. Build that, and rebuild DALIGNER. Then install. I'm eager to hear of positive or negative experiences.

I can't tag a new release of FALCON-integrate until I discuss this with others. Maybe a day or two. I'll post then.

**Comment by pb-cdunn** [comment]
The tip of master for **FALCON-integrate** has all the [latest code](https://github.com/PacificBiosciences/FALCON-integrate/pull/16).

**Comment by aliceb5** [comment]
Thanks for the reply!

I think I might be getting the same error ...   I'm worried perhaps I
didn't download the right version, though.

I downloaded using git clone from here:
https://github.com/PacificBiosciences/FALCON-integrate

My prepare_pdb.sh.log is attached.

Thanks!

On Wed, Nov 4, 2015 at 11:32 PM, Christopher Dunn <notifications@github.com>
wrote:

> The tip of master for *FALCON-integrate* has all the latest code
> <https://github.com/PacificBiosciences/FALCON-integrate/pull/16>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/PacificBiosciences/FALCON/issues/224#issuecomment-153890669>
> .
>



-- 


Alice Dennis
alicebdennis@gmail.com

Postdoctoral Researcher
Institute for Integrative Biology, ETH Zürich & EAWAG
Überlandstrasse 133
P.O. Box 611
8600 Dübendorf, Switzerland

https://adennis5.wordpress.com/

**Comment by pb-cdunn** [comment]
Run this in both `FALCON-integrate` and `FALCON-integreate/FALCON`:

    git rev-parse HEAD

That will tell us the actual versions you are using. My guess is that you need to do this:

    cd FALCON-integrate
    git pull
    git submodule update

**Comment by aliceb5** [comment]
Thanks for the reply.

git rev-parse HEAD gives me this:

From FALCON-integrate
4822a1253632d23078e275162d60a6ba1a01a803

From FALCON
4822a1253632d23078e275162d60a6ba1a01a803

And git pull gives me: "Already up-to-date"


Do these make sense?  I deleted everything and re-installed today...

Thank you!


On Thu, Nov 5, 2015 at 4:45 PM, Christopher Dunn <notifications@github.com>
wrote:

> Run this in both FALCON-integrate and FALCON-integreate/FALCON:
>
> git rev-parse HEAD
>
> That will tell us the actual versions you are using. My guess is that you
> need to do this:
>
> cd FALCON-integrate
> git pull
> git submodule update
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/PacificBiosciences/FALCON/issues/224#issuecomment-154098170>
> .
>



-- 


Alice Dennis
alicebdennis@gmail.com

Postdoctoral Researcher
Institute for Integrative Biology, ETH Zürich & EAWAG
Überlandstrasse 133
P.O. Box 611
8600 Dübendorf, Switzerland

https://adennis5.wordpress.com/

**Comment by pb-cdunn** [comment]
Sorry. I meant `DAZZ_DB`, not `FALCON`. Could you send that one too?

And I don't see any attachment. Could you put the log-file elsewhere? Maybe in a [GitHub gist](https://gist.github.com/)?

**Comment by aliceb5** [comment]
Here it is from DAZZ_DB:
Submodule path 'DAZZ_DB': checked out '993e2bd578fbdf042cf52d41eaccf2cde35c444c'

I should add that Installing FALCON required quite a bit of wrangling on my computing cluster because (1) the virtual environment wasn't allowed, and (2) I didn't have enough space available to me to run Falcon in a permanent directory.  So, with the help of cluster admin, the best solution we could find was to put the install in a temporary (scratch) area of the cluster, and all of the executables are in a single FALCON/bin folder.

Anyway, this seemed to have worked until I hit this snag, and since other people are (or were) seeing this I hope that this install isn't the problem!  

I have pasted my prepare_pdb.sh.log file here:
https://gist.github.com/aliceb5/475b67272318aa687003

Thank you!

**Comment by pb-cdunn** [comment]
> Submodule path 'DAZZ_DB': checked out '993e2bd578fbdf042cf52d41eaccf2cde35c444c'

We don't have that commit in our repo. Could you show the log?

    git log --pretty=fuller 993e2bd578fbdf042cf52d41eaccf2cde35c444c

**Comment by aliceb5** [comment]
Here is the results from the above, run in the DAZZ_DB directory:
https://gist.github.com/aliceb5/bf7f1bc096b70e5f1c76

**Comment by pb-cdunn** [comment]
That's the right commit. Either you haven't built and installed the new fasta2DB yet, or it just doesn't work. I haven't actually tried it yet. flowers9@ seemed to be pretty sure it worked though.

I can debug this next week. If you want something sooner, you might have to look into it yourself.

**Comment by aliceb5** [comment]
Ok, thanks. 

I deleted everything and re-installed, but still get this error. I ended up using fastx_toolkit to wrap my fasta files with lines >8000 bp (just an arbitrary number).

Now I'm getting this error (this is the whole of prepare_pbd.sh.log):
....
trap 'touch /cluster/scratch_xp/public/adennis/FALCON/bin/1-preads_ovl/pdb_build_done.exit' EXIT
+ trap 'touch /cluster/scratch_xp/public/adennis/FALCON/bin/1-preads_ovl/pdb_build_done.exit' EXIT
cd /cluster/scratch_xp/public/adennis/FALCON/bin/1-preads_ovl
+ cd /cluster/scratch_xp/public/adennis/FALCON/bin/1-preads_ovl
hostname
+ hostname
a6032.hpc-net.ethz.ch
date
+ date
Fri Nov  6 11:48:49 CET 2015
fasta2DB -v preads -f/cluster/scratch_xp/public/adennis/FALCON/bin/1-preads_ovl/preads.fofn2
+ fasta2DB -v preads -f/cluster/scratch_xp/public/adennis/FALCON/bin/1-preads_ovl/preads.fofn2
Adding 'all.trimmed.wrap' ...
File all.trimmed.wrap.fasta, Line 177477: Pacbio header line name inconsisten
touch /cluster/scratch_xp/public/adennis/FALCON/bin/1-preads_ovl/pdb_build_done.exit
+ touch /cluster/scratch_xp/public/adennis/FALCON/bin/1-preads_ovl/pdb_build_done.exit
....

But I don't see anything wrong with this line. Is there some magic number that shouldn't be there?  

 sed '177477!d' /cluster/scratch_xp/public/adennis/all.subreads/all.trimmed.wrap.fasta
>m140203_212434_42164_c100608902550000001823099106251492_s1_p0/148/7721_13847 SUBSTR:0,6634

The other lines around it are similarly formatted:
sed '177479!d' /cluster/scratch_xp/public/adennis/all.subreads/all.trimmed.wrap.fasta
>m140203_212434_42164_c100608902550000001823099106251492_s1_p0/208/0_7499 SUBSTR:17,7496

Thanks for any suggestions!

**Comment by pb-cdunn** [comment]
Did fastx_toolkit change the header lines? Compare before and after. Also compare line 177477 with line 1. Post here if interesting.

**Comment by pb-jchin** [comment]
are the `SUBSTR` part of the headers?  `fasta2DB` has a very simple parser for the header

**Comment by aliceb5** [comment]
Yes, the SUBSTR is at the end of every line, preceded by a space. That does seem like it might be a problem!

So I should run these through fasta2DB before running fc_run.py?  If so, do I do that with or without the SUBSTR part?

Thanks

**Comment by pb-cdunn** [comment]
We can modify fasta2DB to allow this string. Please show me line 1 and line 177477.

    head -n 1 x.fasta
    tail -n 177477 x.fasta | head -n  1

Also, please show me line 1 *before* fastax_toolkit.

Post all of that within [triple-backticks](https://help.github.com/articles/github-flavored-markdown/).

**Comment by dshvets** [comment]
Hi I am about to try this again - so has it been updated to the point where all I have to do is just re-download and install all of falcon all over again? 
Thank you!

**Comment by pb-cdunn** [comment]
@dshvets, yes, the latest FALCON-integrate (on `master` branch) should work for you. `SUBSTR` is a different problem which you will not have, with luck.

@aliceb5, if you can post those details on the header lines, please open a new ticket.

**Comment by lauraalazar** [comment]
I am having the same issue described above: 
File filtered_subreads.fasta, Line 24: Fasta line is too long

I cloned falcon 0.4 following the steps in: https://github.com/PacificBiosciences/Bioinformatics-Training/wiki/FALCON-Walkthrough:-Assembling-Lambda-and-Analyzing-Output

I tried to fix my problem by using "FALCON-formatter": https://github.com/zyndagj/FALCON-formatter
but then got the error: File filtered_subreads.fasta, Line 16504703: Pacbio header line name inconsisten

When I do: git checkout master
Everything is up-to-date with 'origin/master'

**Comment by pb-cdunn** [comment]
You need a later version. Try branch `master` or tag `4.2`.
* https://github.com/PacificBiosciences/FALCON-integrate/releases

FALCON-integrate uses **git-submodules**. You will need to update your submodules, as your version of DAZZ_DB is old.

**Comment by lauraalazar** [comment]
`cd FALCON-integrate`
`git rev-parse HEAD`
4e655db1d06d99301b1d75bf1c13efc94a8d66c5

`cd DAZZ_DB`
`git rev-parse HEAD`
ed0b85e6f14dc394fbabd0731187f98e08a79a0a

Thanks!

**Comment by pb-cdunn** [comment]
@lauraalazar,
```
cd FALCON-integrate
git submodule update
```
After that, you DAZZ_DB ref will be `9d43184`. Rebuild, and I think that will solve your problem.

**Comment by lauraalazar** [comment]
Thank you very much, It is now running! 
I realized that my input was also incorrect because I had a single fasta file that I created from the fastq.  I now split the fasta files with FALCON-formatter and after the update of DAZZ_DB it seems to be working (so far).

**Comment by pb-cdunn** [comment]
Which FALCON-formatter do you use?

**Comment by lauraalazar** [comment]
I used this: https://github.com/zyndagj/FALCON-formatter
